sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala

```diff
@@ -1343,17 +1343,41 @@ class HiveDDLSuite
       sql("INSERT INTO t SELECT 2, 'b'")
       checkAnswer(spark.table("t"), Row(9, "x") :: Row(2, "b") :: Nil)
 
-      val e = intercept[AnalysisException] {
-        Seq(1 -> "a").toDF("i", "j").write.format("hive").partitionBy("i").saveAsTable("t2")
-      }
-      assert(e.message.contains("A Create Table As Select (CTAS) statement is not allowed " +
-        "to create a partitioned table using Hive"))
-
       val e2 = intercept[AnalysisException] {
         Seq(1 -> "a").toDF("i", "j").write.format("hive").bucketBy(4, "i").saveAsTable("t2")
       }
       assert(e2.message.contains("Creating bucketed Hive serde table is not supported yet"))
 
+      try {
+        spark.sql("set hive.exec.dynamic.partition.mode=nonstrict")
+        Seq(10 -> "y").toDF("i", "j").write.format("hive").partitionBy("i").saveAsTable("t3")
+        checkAnswer(spark.table("t3"), Row("y", 10) :: Nil)
+        table = spark.sessionState.catalog.getTableMetadata(TableIdentifier("t3"))
+        var partitionSchema = table.partitionSchema
+        assert(partitionSchema.size == 1 && partitionSchema.fields(0).name == "i" &&
+          partitionSchema.fields(0).dataType == IntegerType)
+
+        Seq(11 -> "z").toDF("i", "j").write.mode("overwrite").format("hive")
+          .partitionBy("j").saveAsTable("t3")
+        checkAnswer(spark.table("t3"), Row(11, "z") :: Nil)
+        table = spark.sessionState.catalog.getTableMetadata(TableIdentifier("t3"))
+        partitionSchema = table.partitionSchema
+        assert(partitionSchema.size == 1 && partitionSchema.fields(0).name == "j" &&
+          partitionSchema.fields(0).dataType == StringType)
+
+        Seq((1, 2, 3)).toDF("i", "j", "k").write.mode("overwrite").format("hive")
+          .partitionBy("k", "j").saveAsTable("t3")
+        table = spark.sessionState.catalog.getTableMetadata(TableIdentifier("t3"))
+        checkAnswer(spark.table("t3"), Row(1, 3, 2) :: Nil)
+
+        Seq((1, 2, 3)).toDF("i", "j", "k").write.mode("overwrite").format("hive")
+          .partitionBy("j", "k").saveAsTable("t3")
+        table = spark.sessionState.catalog.getTableMetadata(TableIdentifier("t3"))
+        checkAnswer(spark.table("t3"), Row(1, 2, 3) :: Nil)
+      } finally {
+        spark.sql("set hive.exec.dynamic.partition.mode=strict")
+      }
+
```

Comment:
I think this test case is a bit fat, maybe we can split it into two or three smaller ones? e.g.:

```scala
  test("create hive serde table with DataFrameWriter.saveAsTable - basic") ...
  test("create hive serde table with DataFrameWriter.saveAsTable - overwrite and append") ...
  test("create hive serde table with DataFrameWriter.saveAsTable - partitioned") ...
```